### PR TITLE
Maintain order book

### PR DESCRIPTION
**Pull request by Chralt98**
### What does it do?

Fixes https://github.com/zeitgeistpm/zeitgeist/issues/1116

Fixes https://github.com/zeitgeistpm/zeitgeist/issues/1115

Fixes https://github.com/zeitgeistpm/zeitgeist/issues/1117

### CODE REVIEW

File: zrml/orderbook/src/tests.rs [hunk 1]
```diff
@@ -80,12 +116,410 @@ fn fill_order_fails_with_wrong_scoring_rule(scoring_rule: ScoringRule) {
...
+            RuntimeOrigin::signed(ALICE),
+            market_id,
+            maker_asset,
+            10 * BASE,
+            taker_asset,
+            taker_amount,
+        ));
+
+        let bob_free_taker_asset = AssetManager::free_balance(taker_asset, &BOB);
+        assert_eq!(bob_free_taker_asset, 0);
```
Comment (maltekliemann) [inline zrml/orderbook/src/tests.rs:295]: Ideally, Bob would own one unit less than required instead of nothing.

File: zrml/orderbook/src/tests.rs [hunk 2]
```diff
@@ -80,12 +116,410 @@ fn fill_order_fails_with_wrong_scoring_rule(scoring_rule: ScoringRule) {
...
+            maker_asset,
+            10 * BASE,
+            taker_asset,
+            taker_amount,
+        ));
+
+        AssetManager::deposit(taker_asset, &BOB, taker_amount).unwrap();
+
+        assert_noop!(
+            Orderbook::fill_order(RuntimeOrigin::signed(BOB), order_id, Some(taker_amount - 1)),
```
Comment (maltekliemann) [inline zrml/orderbook/src/tests.rs:329]:
```suggestion
            Orderbook::fill_order(RuntimeOrigin::signed(BOB), order_id, Some(taker_amount - AssetManager::minimum_balance(taker_asset) + 1)),

File: zrml/orderbook/src/tests.rs [hunk 3]
```diff
@@ -80,12 +116,410 @@ fn fill_order_fails_with_wrong_scoring_rule(scoring_rule: ScoringRule) {
...
+        let market_id = 0u128;
+        let market = market_mock::<Runtime>();
+        Markets::<Runtime>::insert(market_id, market.clone());
+
+        let maker_asset = market.base_asset;
+        let taker_asset = Asset::CategoricalOutcome(0, 2);
+
+        let order_id = 0u128;
+        let taker_amount = 25 * BASE;
+
```
Comment (maltekliemann) [inline zrml/orderbook/src/tests.rs:377]:
```suggestion
        let maker_amount = 10 * BASE;

```

File: zrml/orderbook/src/tests.rs [hunk 4]
```diff
@@ -80,12 +116,410 @@ fn fill_order_fails_with_wrong_scoring_rule(scoring_rule: ScoringRule) {
...
+        assert_ok!(Orderbook::place_order(
+            RuntimeOrigin::signed(ALICE),
+            market_id,
+            maker_asset,
+            10 * BASE,
```
Comment (maltekliemann) [inline zrml/orderbook/src/tests.rs:382]:
```suggestion
            maker_amount,
```

File: zrml/orderbook/src/tests.rs [hunk 5]
```diff
@@ -80,12 +116,410 @@ fn fill_order_fails_with_wrong_scoring_rule(scoring_rule: ScoringRule) {
...
+        AssetManager::deposit(taker_asset, &BOB, taker_amount).unwrap();
+
+        let order = <Orders<Runtime>>::get(order_id).unwrap();
+        assert_eq!(
+            order,
+            Order {
+                market_id,
+                maker: ALICE,
+                maker_asset,
+                maker_amount: 10 * BASE,
```
Comment (maltekliemann) [inline zrml/orderbook/src/tests.rs:396]:
```suggestion
                maker_amount,
```

File: zrml/orderbook/src/tests.rs [hunk 6]
```diff
@@ -80,12 +116,410 @@ fn fill_order_fails_with_wrong_scoring_rule(scoring_rule: ScoringRule) {
...
+                taker_asset,
+                taker_amount: 25 * BASE,
```
Comment (maltekliemann) [inline zrml/orderbook/src/tests.rs:398]:
```suggestion
                taker_amount,
```

File: zrml/orderbook/src/tests.rs [hunk 7]
```diff
@@ -80,12 +116,410 @@ fn fill_order_fails_with_wrong_scoring_rule(scoring_rule: ScoringRule) {
...
+            }
+        );
+
+        assert_ok!(Orderbook::fill_order(
+            RuntimeOrigin::signed(BOB),
+            order_id,
+            Some(taker_amount / 2)
+        ));
+
+        let order = <Orders<Runtime>>::get(order_id).unwrap();
```
Comment (maltekliemann) [inline zrml/orderbook/src/tests.rs:408]: `Order::<Runtime>::foo`

File: zrml/orderbook/src/tests.rs [hunk 8]
```diff
@@ -80,12 +116,410 @@ fn fill_order_fails_with_wrong_scoring_rule(scoring_rule: ScoringRule) {
...
+
+        assert_eq!(
```
Comment (maltekliemann) [inline zrml/orderbook/src/tests.rs:410]: I'd say this test should also verify that the amounts are correctly changed.

File: zrml/orderbook/src/tests.rs [hunk 9]
```diff
@@ -80,12 +116,410 @@ fn fill_order_fails_with_wrong_scoring_rule(scoring_rule: ScoringRule) {
...
+            taker_amount,
+        ));
+
+        AssetManager::deposit(taker_asset, &BOB, taker_amount).unwrap();
+
+        assert_ok!(Orderbook::fill_order(RuntimeOrigin::signed(BOB), order_id, None));
+
+        assert!(<Orders<Runtime>>::get(order_id).is_none());
+    });
+}
```
Comment (maltekliemann) [inline zrml/orderbook/src/tests.rs:363]: Why isn't this part of `it_fills_order_fully_maker_outcome_asset`?

File: zrml/orderbook/src/tests.rs [hunk 10]
```diff
@@ -107,367 +541,506 @@ fn it_places_orders() {
...
+        assert_eq!(market_creator_balance_after - market_creator_balance_before, taker_fees);
+
+        let reserved_bob = AssetManager::reserved_balance(maker_asset, &BOB);
         assert_eq!(reserved_bob, 0);
 
         System::assert_last_event(
             Event::<Runtime>::OrderFilled {
                 order_id,
                 maker: BOB,
                 taker: ALICE,
```
Comment (maltekliemann) [inline zrml/orderbook/src/tests.rs:629]: This event doesn't include which part of the maker amount was traded. I guess we could read that off the the transaction data. Is that standard practice, or should we add the other amounts too, @saboonikhil?

File: zrml/orderbook/src/tests.rs [hunk 11]
```diff
@@ -107,367 +541,506 @@ fn it_places_orders() {
...
         );
 
-        let alice_bal = <Balances as Currency<AccountIdTest>>::free_balance(&ALICE);
-        let alice_shares = Tokens::free_balance(outcome_asset, &ALICE);
-        assert_eq!(alice_bal, BASE - 50);
-        assert_eq!(alice_shares, 10);
+        let alice_maker_asset_free = AssetManager::free_balance(taker_asset, &ALICE);
+        let alice_taker_asset_free = AssetManager::free_balance(maker_asset, &ALICE);
+        assert_eq!(alice_maker_asset_free, INITIAL_BALANCE);
+
```
Comment (maltekliemann) [inline zrml/orderbook/src/tests.rs:640]:
```suggestion
```

File: zrml/orderbook/src/tests.rs [hunk 12]
```diff
@@ -107,367 +541,506 @@ fn it_places_orders() {
...
-        let reserved_bob = Tokens::reserved_balance(outcome_asset, &BOB);
-        assert_eq!(reserved_bob, 860);
+        let reserved_bob = AssetManager::reserved_balance(maker_asset, &BOB);
+        assert_eq!(reserved_bob, filled_maker_amount);
     });
 }
 
 #[test]
-fn it_removes_orders() {
+fn it_removes_order() {
```
Comment (maltekliemann) [inline zrml/orderbook/src/tests.rs:899]: This test seems to be mixing multiple tests (success and failure). I think we should split this.

File: zrml/orderbook/src/mock.rs [hunk 13]
```diff
@@ -38,6 +43,41 @@ use zeitgeist_primitives::{
...
+
+    fn distribute(
+        _market_id: Self::MarketId,
+        asset: Self::Asset,
+        account: &Self::AccountId,
+        amount: Self::Balance,
+    ) -> Self::Balance {
+        let fees = Self::get_fee(_market_id, amount);
+        let _ = T::AssetManager::transfer(asset, account, &F::get(), fees);
+        fees
```
Comment (maltekliemann) [inline zrml/orderbook/src/mock.rs:77]: Looking at this now, this implementation is wrong, isn't it? Because if the transfer fails, then the return value should be zero.

```suggestion
        match T::AssetManager::transfer(asset, account, &F::get(), fees) {
            Ok(_) => fees,
            Err(_) => Zero::zero(),
        }
```

**Review by maltekliemann — CHANGES_REQUESTED** [review]
Don't forget the migration!

File: runtime/common/src/fees.rs [hunk 14]
```diff
@@ -263,14 +267,20 @@ macro_rules! impl_market_creator_fees {
         }
 
         impl MarketCreatorFee {
+            fn do_get_fee(market_id: MarketId, amount: Balance) -> Result<Balance, DispatchError> {
+                let market = MarketCommons::market(&market_id)?; // Should never fail
+                let fee_amount = market.creator_fee.mul_floor(amount);
+                Ok(fee_amount)
+            }
+
```
[the remark this answers is not in the document]
Comment (maltekliemann) [reply runtime/common/src/fees.rs:275]: > Are you satisfied if we use unwrap_or_default() in both instances? Balance implements the default trait using Zero.

Seems fine to me. After all, the function always return the amount of fees deducted. If the transfer fails, no fees change hands, so the function should return zero.

> I don't think your suggestion solves the problem what needs to happen if do_distribute errors.

What's the problem?

[the remark this answers is not in the document]
Comment (maltekliemann) [reply runtime/common/src/fees.rs:275]: Ah, so the problem is that `get_fee` returns the wrong value if `distribute` errors. That's a good argument for always using the return value of `distribute` instead of `get_fee`. If you want to implement `get_fee` correctly, I guess you need to ensure that the transfer will actually go through or return zero.

File: zrml/orderbook/src/benchmarks.rs [hunk 15]
```diff
@@ -50,71 +48,69 @@ fn generate_funded_account<T: Config>(seed: Option<u32>) -> Result<T::AccountId,
...
-
-    fill_order_bid {
-        let caller = generate_funded_account::<T>(None)?;
-        let (_, _, order_id) = place_order::<T>(OrderSide::Bid, Some(0))?;
+    fill_order {
+        let taker_asset = Asset::CategoricalOutcome::<MarketIdOf<T>>(0u32.into(), 0);
+        let caller = generate_funded_account::<T>(None, taker_asset)?;
+        let (_, market_id, order_id) = place_default_order::<T>(Some(0), taker_asset)?;
+        let maker_asset = T::MarketCommons::market(&market_id).unwrap().base_asset;
+        let caller = generate_funded_account::<T>(None, maker_asset)?;
```
[the remark this answers is not in the document]
Comment (maltekliemann) [reply zrml/orderbook/src/benchmarks.rs]: It's not a major problem. But even now it's really weird design, because you're now handling two markets IDs and solving the problem by just throwing one of them away.

File: zrml/orderbook/src/tests.rs [hunk 16]
```diff
@@ -80,12 +116,410 @@ fn fill_order_fails_with_wrong_scoring_rule(scoring_rule: ScoringRule) {
...
 
+#[test_case(MarketStatus::Proposed; "proposed")]
+#[test_case(MarketStatus::Suspended; "suspended")]
+#[test_case(MarketStatus::Closed; "closed")]
+#[test_case(MarketStatus::CollectingSubsidy; "collecting subsidy")]
+#[test_case(MarketStatus::InsufficientSubsidy; "insufficient subsidy")]
+#[test_case(MarketStatus::Reported; "reported")]
+#[test_case(MarketStatus::Disputed; "disputed")]
+#[test_case(MarketStatus::Resolved; "resolved")]
+fn fill_order_fails_if_market_status_not_active(status: MarketStatus) {
```
[the remark this answers is not in the document]
Comment (maltekliemann) [reply zrml/orderbook/src/tests.rs:136]: I think I put this comment in the wrong place. Sorry. Let me try again.

What I mean is that `remove_order` doesn't require that the market have scoring rule `OrderBook`, while `fill_order` does. In both cases you could argue that it is not necessary to have this requirement because `place_order` won't allow placing orders for markets with the wrong scoring rule to begin with. That's why you have to use `mutate_market` to create a scenario which can't happen in the first place just to test this error.

I think both approaches are valid, but we should stick with one approach for both functions.

File: zrml/orderbook/src/mock.rs [hunk 17]
```diff
@@ -38,6 +43,44 @@ use zeitgeist_primitives::{
...
+        account: &Self::AccountId,
+        amount: Self::Balance,
+    ) -> Self::Balance {
+        let fees = Self::get_fee(_market_id, amount);
+        match T::AssetManager::transfer(asset, account, &F::get(), fees) {
+            Ok(_) => fees,
+            Err(_) => Zero::zero(),
+        };
+        fees
+    }
```
Comment (maltekliemann) [inline zrml/orderbook/src/mock.rs:81]:
```suggestion
        match T::AssetManager::transfer(asset, account, &F::get(), fees) {
            Ok(_) => fees,
            Err(_) => Zero::zero(),
        }
```

File: zrml/orderbook/src/tests.rs [hunk 18]
```diff
@@ -45,33 +48,69 @@ fn place_order_fails_with_wrong_scoring_rule(scoring_rule: ScoringRule) {
...
+                market_id,
+                market.base_asset,
+                10 * BASE,
+                Asset::CategoricalOutcome(0, 2),
+                25 * BASE,
+            ),
+            Error::<Runtime>::MarketIsNotActive
+        );
+    });
+}
```
Comment (maltekliemann) [inline zrml/orderbook/src/tests.rs:91]:
```suggestion
}

```

File: runtime/common/src/fees.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (maltekliemann) [reply runtime/common/src/fees.rs:275]: You can cherry-pick from this branch: https://github.com/zeitgeistpm/zeitgeist/tree/mkl-orderbook-abstraction

I'd say `get_fee` must either be rewritten as described above _or_ market as only to be used for testing purposes.

File: zrml/orderbook/src/lib.rs [hunk 19]
```diff
@@ -160,199 +174,307 @@ mod pallet {
...
+                    order_data.market_id,
+                    // maker_fill is the amount what the maker wants to have (base asset from taker)
+                    maker_fill,
+                );
+                // base asset amount from taker minus fees to the maker (maker_fill)
+                maker_fill.checked_sub(&external_fees).ok_or(ArithmeticError::Underflow)?
+            };
+
+            Ok(maybe_adjusted_maker_fill)
+        }
```
[the remark this answers is not in the document]
Comment (maltekliemann) [reply zrml/orderbook/src/lib.rs:409]: You can cherry-pick from this branch: https://github.com/zeitgeistpm/zeitgeist/tree/mkl-orderbook-abstraction

[the remark this answers is not in the document]
Comment (maltekliemann) [reply zrml/orderbook/src/lib.rs:409]: Note: Given the fact that `get_fee` currently doesn't always do the right thing (and it may be difficult to ensure that it does), it probably shouldn't be used and `distribute` should be used instead.

File: zrml/orderbook/src/tests.rs [hunk 20]
```diff
@@ -107,367 +534,536 @@ fn it_places_orders() {
...
+        ));
+
+        let order_id = 0u128;
+
         assert_noop!(
             Orderbook::remove_order(RuntimeOrigin::signed(BOB), order_id),
             Error::<Runtime>::NotOrderCreator,
         );
+    });
+}
```
Comment (maltekliemann) [inline zrml/orderbook/src/tests.rs:968]: This is already covered by `remove_order_fails_if_not_order_creator`.

```suggestion
```

**Review by maltekliemann — CHANGES_REQUESTED** [review]

File: zrml/orderbook/src/lib.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/orderbook/src/lib.rs]: I only see the usage in the tests. I think about how we can reach the same thing in a better way.

File: runtime/common/src/fees.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (Chralt98) [reply runtime/common/src/fees.rs]: 102a742142560efc91718c967f01c95b95cfbdc7

File: zrml/orderbook/src/benchmarks.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/orderbook/src/benchmarks.rs]: I admit, I don't really give a high attention to the design of benchmarks as soon as they are quick to write. What do you propose us to do now. Can you suggest something or should we just ignore this issue?

File: zrml/orderbook/src/tests.rs (hunk 20 above)
Comment (Chralt98) [reply zrml/orderbook/src/tests.rs]: 7f88460e725344dfb180576fb785fa9a737a47c8

File: zrml/orderbook/src/tests.rs (hunk 18 above)
Comment (Chralt98) [reply zrml/orderbook/src/tests.rs:90]: 7f88460e725344dfb180576fb785fa9a737a47c8

File: zrml/orderbook/src/mock.rs (hunk 17 above)
Comment (Chralt98) [reply zrml/orderbook/src/mock.rs]: 7f88460e725344dfb180576fb785fa9a737a47c8

File: zrml/orderbook/src/tests.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/orderbook/src/tests.rs:136]: I still don't get it fully yet. 

So, should we remove the code testing a scenario that wouldn't occur? Or should we add a scoring rule check for remove_order?

[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/orderbook/src/tests.rs:136]: I would also just remove the check from `fill_order`, because `place_order` prevents this. However I would add a debug_assert in fill order.

[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/orderbook/src/tests.rs:100]: 53ef53e1197067a5642e8dd3b7d550b17289555a

**Comment by Chralt98** [comment]
I will do the migration in this PR shortly.

File: zrml/orderbook/src/migrations.rs
Comment (Chralt98) [inline zrml/orderbook/src/migrations.rs:1]: I did test the try-runtime run on main and test net. Unfortunately we have no examples other than my written migration tests here.. Because there is no order with the old structure open at the moment.

File: zrml/orderbook/src/migrations.rs [hunk 21]
```diff
@@ -0,0 +1,336 @@
...
+            // assert old fields
+            assert_eq!(old_order.market_id, new_order.market_id);
+            assert_eq!(old_order.maker, new_order.maker);
+            // assert new fields
+            let reserved = T::AssetManager::reserved_balance_named(
+                &OrderbookPallet::<T>::reserve_id(),
+                new_order.maker_asset,
+                &new_order.maker,
+            );
+            assert!(reserved > BalanceOf::<T>::zero());
```
Comment (maltekliemann) [inline zrml/orderbook/src/migrations.rs:176]:
```suggestion
```

Comment (maltekliemann) [reply zrml/orderbook/src/migrations.rs:176]: This is covered by the next `assert_eq`

File: zrml/orderbook/src/migrations.rs [hunk 22]
```diff
@@ -0,0 +1,336 @@
...
+        });
+    }
+
+    fn set_up_version() {
+        StorageVersion::new(ORDER_BOOK_REQUIRED_STORAGE_VERSION).put::<OrderbookPallet<Runtime>>();
+    }
+
+    fn construct_old_new_tuple() -> (Vec<OldOrderOf<Runtime>>, Vec<OrderOf<Runtime>>) {
+        let market_id_0 = 0;
+        let outcome_asset_amoun_0 = 42000;
```
Comment (maltekliemann) [inline zrml/orderbook/src/migrations.rs:278]:
```suggestion
        let outcome_asset_amount_0 = 42000;
```

Etc.

File: zrml/orderbook/src/migrations.rs [hunk 23]
```diff
@@ -0,0 +1,336 @@
...
+            maker: 1,
+            // the maker reserved the base asset for order side bid
+            maker_asset: Asset::Ztg,
+            maker_amount: base_asset_amount_0,
+            taker_asset: Asset::CategoricalOutcome(market_id_0, 0u16),
+            taker_amount: outcome_asset_amoun_0,
+        };
+
+        let market_id_1 = 1;
+        let outcome_asset_amoun_1 = 42000;
```
Comment (maltekliemann) [inline zrml/orderbook/src/migrations.rs:300]: same here

File: zrml/orderbook/src/benchmarks.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (maltekliemann) [reply zrml/orderbook/src/benchmarks.rs]: No just leave it

File: zrml/orderbook/src/migrations.rs [hunk 24]
```diff
@@ -0,0 +1,336 @@
...
+            );
+            TranslateOrderStructure::<Runtime>::on_runtime_upgrade();
+
+            let actual = crate::Orders::<Runtime>::get(0).unwrap();
+            assert_eq!(actual, new_orders[0]);
+
+            let actual = crate::Orders::<Runtime>::get(1).unwrap();
+            assert_eq!(actual, new_orders[1]);
+        });
+    }
```
Comment (maltekliemann) [inline zrml/orderbook/src/migrations.rs:270]: This is a duplicate of the previous test.

**Review by maltekliemann — CHANGES_REQUESTED** [review]

File: zrml/orderbook/src/migrations.rs (hunk 24 above)
Comment (maltekliemann) [reply zrml/orderbook/src/migrations.rs:270]: Nevermind, I can't read.

File: zrml/orderbook/README.md [hunk 25]
```diff
@@ -1,3 +1,46 @@
...
+asset for outcome assets and vice versa.
+
+## Overview
+
+The order book can be set as a market's scoring rule. It allows to place,
+partially or fully fill and remove orders.
+
+## Terminology
+
+- `maker_partial_fill`: The partial amount of what the maker wants to fill.
```
Comment (Robiquet) [inline zrml/orderbook/README.md:13]: How can a maker fill an order? Wouldn't that make them a taker?

Comment (Chralt98) [reply zrml/orderbook/README.md:13]: Haha, good point @Robiquet .. Malte already pointed this out extensively. Good that you found another instance for this.

File: zrml/neo-swaps/src/mock.rs [hunk 26]
```diff
@@ -140,12 +140,9 @@ where
...
-            amount.saturated_into(),
-            EXTERNAL_FEES.saturated_into(),
-        )
-        .unwrap()
-        .saturated_into();
+        let fees = zeitgeist_primitives::math::fixed::bmul(amount.saturated_into(), EXTERNAL_FEES)
+            .unwrap()
+            .saturated_into();
         let _ = T::MultiCurrency::transfer(asset, account, &F::get(), fees);
         fees
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/neo-swaps/src/mock.rs]: 1b3fc077ce0661258af610ee8691a14adb712c06

File: zrml/orderbook/src/benchmarks.rs [hunk 27]
```diff
@@ -24,97 +24,87 @@
...
 }
 
-// Creates an account and gives it asset and currency. `seed` specifies the account seed,
-// None will return a whitelisted account
-// Returns `account`, `asset`, `outcome_asset_amount` and `base_asset_amount`
 fn order_common_parameters<T: Config>(
     seed: Option<u32>,
 ) -> Result<
-    (T::AccountId, Asset<MarketIdOf<T>>, BalanceOf<T>, BalanceOf<T>, MarketIdOf<T>),
+    (MarketIdOf<T>, T::AccountId, Asset<MarketIdOf<T>>, BalanceOf<T>, BalanceOf<T>),
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/orderbook/src/benchmarks.rs:47]: I would say I changed the order, because the market_id seemed to be the most fundamental value.

File: zrml/orderbook/src/migrations.rs (hunk 21 above)
Comment (Chralt98) [reply zrml/orderbook/src/migrations.rs]: 186b1d2ac14187f85bb95f25c37019c501fb6122

File: zrml/orderbook/src/migrations.rs (hunk 22 above)
Comment (Chralt98) [reply zrml/orderbook/src/migrations.rs]: 41897a068921646edc68920f363fd7f048407350

File: zrml/orderbook/src/migrations.rs (hunk 23 above)
Comment (Chralt98) [reply zrml/orderbook/src/migrations.rs]: 623e08f62b1864b2b7a60cfb30e3bb657ebdbb57

File: zrml/orderbook/README.md (hunk 25 above)
Comment (Chralt98) [reply zrml/orderbook/README.md]: a0e0f21bb21a7e1044cfaa4ae95d23329a30a394

File: zrml/orderbook/src/tests.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/orderbook/src/tests.rs:100]: 011547635e993322cf4be2e5bde34c6a7a7b0790

File: zrml/orderbook/src/migrations.rs
Comment (Chralt98) [reply zrml/orderbook/src/migrations.rs:1]: I created some orders and used try-runtime again. I had to adjust the try-runtime post-upgrade function. After this everything worked as expected! 

209adc629f122f6b33a016a5557ff8d248bd951f

File: zrml/orderbook/src/migrations.rs [hunk 28]
```diff
@@ -0,0 +1,335 @@
...
+                new_orders.clone(),
+            );
+            TranslateOrderStructure::<Runtime>::on_runtime_upgrade();
+
+            let actual = crate::Orders::<Runtime>::get(0).unwrap();
+            assert_eq!(actual, new_orders[0]);
+
+            let actual = crate::Orders::<Runtime>::get(1).unwrap();
+            assert_eq!(actual, new_orders[1]);
+        });
```
[the remark this answers is not in the document]
Comment (maltekliemann) [reply zrml/orderbook/src/migrations.rs:268]: Funny thing is - this is very difficult to figure out because `cargo test` eats `log::info!`, even with `--nocapture` enabled.

[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/orderbook/src/migrations.rs]: c19a2e5215b4f73d44e5c60c002599081bd78c42

File: zrml/orderbook/src/migrations.rs [hunk 29]
```diff
@@ -251,20 +253,14 @@ mod tests {
...
     fn on_runtime_upgrade_is_noop_if_versions_are_not_correct() {
         ExtBuilder::default().build().execute_with(|| {
-            // Don't set up chain to signal that storage is already up to date.
-            let (_, new_orders) = construct_old_new_tuple();
-            populate_test_data::<Twox64Concat, OrderId, OrderOf<Runtime>>(
-                ORDER_BOOK,
-                ORDERS,
-                new_orders.clone(),
-            );
-            TranslateOrderStructure::<Runtime>::on_runtime_upgrade();
```
Comment (maltekliemann) [inline zrml/orderbook/src/migrations.rs]: You should leave the orders in to ensure they don't get messed up by the migration.

**Review by maltekliemann — APPROVED** [review]

File: zrml/orderbook/src/migrations.rs (hunk 29 above)
Comment (Chralt98) [reply zrml/orderbook/src/migrations.rs]: I thought about this as I were making this change but came to the conclusion, that the `StorageVersion` would be changed in any case. So that's why I didn't care about the orders. If the storage root is the same before and after the migration and I know that the `StorageVersion` is changed in any case, I know that the orders are also not changed.

**Review by maltekliemann — APPROVED** [review]

File: zrml/orderbook/src/lib.rs [hunk 30]
```diff
@@ -164,181 +178,274 @@ mod pallet {
...
-                OrderSide::Ask => order_data.base_asset_amount,
-            };
-            let maker_fill = maker_partial_fill.unwrap_or(makers_requested_total);
-            ensure!(!maker_fill.is_zero(), Error::<T>::AmountIsZero);
-            ensure!(maker_fill <= makers_requested_total, Error::<T>::AmountTooHighForOrder);
+            let maker_fill = maker_partial_fill.unwrap_or(order_data.taker_amount);
+            ensure!(
+                maker_fill >= T::AssetManager::minimum_balance(order_data.taker_asset),
+                Error::<T>::BelowMinimumBalance
+            );
```
Comment (sea212) [inline zrml/orderbook/src/lib.rs:403]: Does it mean an order gets stuck once the remaining balance the maker wants is below the minimum balance? In that case I recommend moving this check to when the actual transferal of the funds happens, because in most cases the maker will already have some balance for the requested token and fulfill this requirement.

File: zrml/orderbook/src/lib.rs [hunk 31]
```diff
@@ -164,181 +178,274 @@ mod pallet {
...
+            ensure!(maker_fill <= order_data.taker_amount, Error::<T>::AmountTooHighForOrder);
```
Comment (sea212) [inline zrml/orderbook/src/lib.rs:404]: Just as a side note, I find the naming of the maker and taker information very confusing, as it depends on the perspective that is being looked from. I would recommend sticking to an absolute view, for example `taker_fill` and `taker_amount` always refer to the token that the maker wants.

Comment (sea212) [reply zrml/orderbook/src/lib.rs:404]: This occurs in multiple places in this file.

File: zrml/orderbook/src/lib.rs [hunk 32]
```diff
@@ -164,181 +178,274 @@ mod pallet {
...
 
+            // clone required because of mutable borrow of order_data below
```
Comment (sea212) [inline zrml/orderbook/src/lib.rs:406]:
```suggestion
```

File: zrml/orderbook/README.md [hunk 33]
```diff
@@ -1,3 +1,46 @@
...
+    are charged from the taker after the transaction is executed. In particular,
+    the maker still receives the full amount of outcome tokens.
+
+## Interface
+
+### Dispatches
+
+#### Public Dispatches
+
+- `remove_order`: Allows a user to remove their order from the order book.
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/orderbook/README.md:44]: ChatGPT: In this sentence, "user" is a singular noun, so you should use "a user" rather than "an user." "An" is used before words that begin with a vowel sound, while "a" is used before words that begin with a consonant sound. Since "user" begins with a consonant sound ("yoo"), you should use "a user."

File: zrml/orderbook/src/lib.rs (hunk 30 above)
Comment (Chralt98) [reply zrml/orderbook/src/lib.rs:403]: It just means that a taker, who uses `fill_order` is not allowed to partially fill an order to the extend that there is an amount missing that is below the existential deposit. We ensure somewhere else in the code that the `taker_amount` stays above the existential deposit.

Comment (Chralt98) [reply zrml/orderbook/src/lib.rs:403]: [Here](https://github.com/zeitgeistpm/zeitgeist/blob/7052aefccd264f15d749405764777e9f2e303386/zrml/orderbook/src/lib.rs#L493-L496) we check that `taker_amount` is above the existential deposit. [Here](https://github.com/zeitgeistpm/zeitgeist/blob/7052aefccd264f15d749405764777e9f2e303386/zrml/orderbook/src/lib.rs#L311) we ensure the `taker_amount` is also above the existential deposit in a (partial) fill.

File: zrml/orderbook/src/lib.rs (hunk 31 above)
Comment (Chralt98) [reply zrml/orderbook/src/lib.rs:404]: I don't really understand how to achieve this. I know it's at first hard to wrap the head around it. 

```
The `maker_partial_fill` is the partial amount of what the maker wants to get filled.
```

File: zrml/orderbook/src/lib.rs [hunk 34]
```diff
@@ -164,181 +178,274 @@ mod pallet {
...
+
+        fn do_remove_order(order_id: OrderId, who: AccountIdOf<T>) -> DispatchResult {
+            let order_data = <Orders<T>>::get(order_id).ok_or(Error::<T>::OrderDoesNotExist)?;
+            debug_assert!(
+                {
+                    let market = T::MarketCommons::market(&order_data.market_id)?;
+                    market.scoring_rule == ScoringRule::Orderbook
+                },
+                "The call to place_order already ensured the scoring rule order book."
+            );
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/orderbook/src/lib.rs:323]: Ok, I generally agree that I should reconsider my `debug_assert` usage. I just thought it would be a good way to tell the reader of the code that this assumption is true. `debug_assert` is also an excellent way to have no overhead in the production runtime, since its logic is not included in those builds. We also discussed exactly this code snippet already in the review from Malte. Before using `debug_assert`, I used a normal `ensure!` statement, but Malte suggested that we shouldn't include checks that wouldn't ever happen in production (because we ensure each order created does have the scoring rule order book). Your suggestion comes with the drawback that we would need to runtime overhead in production builds that we wouldn't necessarily need. So at the end, I am in favour of deleting this check overall. What would you do @maltekliemann ?

[the remark this answers is not in the document]
Comment (maltekliemann) [reply zrml/orderbook/src/lib.rs:323]: > but Malte suggested that we shouldn't include checks that wouldn't ever happen in production

I think (hope) that I suggested that you be consistent in using or not using this pattern. In terms of software design, leaving the check out is preferred, in terms of defensive programming, leaving the check in is preferred. Either way is fine.

> The runtime review from SRLabs also expressed that debug_assert in production is a bad practice and should be avoided.

When/if SRLabs publishes a best practices, I'm happing to go over our style guide with you guys and incorporate these. I think there are some other things that we may want to add there.

[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/orderbook/src/lib.rs:323]: > I'm happing to go over our style guide with you guys and incorporate these. I think there are some other things that we may want to add there.

Happy to apply best practices and agree on that.

[the remark this answers is not in the document]
Comment (Chralt98) [reply zrml/orderbook/src/lib.rs]: 537725ec679acb73e5340c7de68ff005f9d3bee3

File: zrml/orderbook/README.md (hunk 33 above)
[the remark this answers is not in the document]
Comment (sea212) [reply zrml/orderbook/README.md:44]: Interesting. Learned something new today :smile:

File: zrml/orderbook/src/lib.rs (hunk 31 above)
Comment (sea212) [reply zrml/orderbook/src/lib.rs:397]: I would just always use `maker` if it affects the maker token and always use `taker` if it affects the taker token.

**Review by sea212 — APPROVED** [review]
